rustls-post-quantum/src/lib.rs

```diff
@@ -43,214 +43,59 @@
-                .map_err(|_| INVALID_KEY_SHARE)?,
-        );
-        Ok(SharedSecret::from(&combined.0[..]))
-    }
-
-    /// Allow the X25519 computation to be offered and selected separately.
-    fn hybrid_component(&self) -> Option<(NamedGroup, &[u8])> {
-        Some((self.x25519.group(), self.x25519.pub_key()))
-    }
-
-    fn complete_hybrid_component(
-        self: Box<Self>,
-        peer_pub_key: &[u8],
-    ) -> Result<SharedSecret, Error> {
-        self.x25519.complete(peer_pub_key)
-    }
-
-    fn pub_key(&self) -> &[u8] {
-        &self.combined_pub_key
-    }
-
-    fn ffdhe_group(&self) -> Option<FfdheGroup<'static>> {
-        None
-    }
-
-    fn group(&self) -> NamedGroup {
-        NamedGroup::X25519MLKEM768
-    }
-}
-
-struct ReceivedShare<'a> {
-    ml_kem: &'a [u8],
-    x25519: &'a [u8],
-}
-
-impl<'a> ReceivedShare<'a> {
-    fn new(buf: &'a [u8]) -> Option<ReceivedShare<'a>> {
-        if buf.len() != COMBINED_PUBKEY_LEN {
-            return None;
-        }
-
-        let (ml_kem, x25519) = buf.split_at(MLKEM768_ENCAP_LEN);
-        Some(ReceivedShare { ml_kem, x25519 })
-    }
-}
-
-struct ReceivedCiphertext<'a> {
-    ml_kem: &'a [u8],
-    x25519: &'a [u8],
-}
-
-impl<'a> ReceivedCiphertext<'a> {
-    fn new(buf: &'a [u8]) -> Option<ReceivedCiphertext<'a>> {
-        if buf.len() != COMBINED_CIPHERTEXT_LEN {
-            return None;
-        }
-
-        let (ml_kem, x25519) = buf.split_at(MLKEM768_CIPHERTEXT_LEN);
-        Some(ReceivedCiphertext { ml_kem, x25519 })
-    }
-}
-
-struct CombinedSecret([u8; COMBINED_SHARED_SECRET_LEN]);
-
-impl CombinedSecret {
-    fn combine(x25519: SharedSecret, ml_kem: kem::SharedSecret) -> Self {
-        let mut out = CombinedSecret([0u8; COMBINED_SHARED_SECRET_LEN]);
-        out.0[..MLKEM768_SECRET_LEN].copy_from_slice(ml_kem.as_ref());
-        out.0[MLKEM768_SECRET_LEN..].copy_from_slice(x25519.secret_bytes());
-        out
-    }
-}
-
-struct CombinedShare(Vec<u8>);
+pub static X25519MLKEM768: &dyn SupportedKxGroup = &hybrid::Hybrid {
+    classical: kx_group::X25519,
+    post_quantum: MLKEM768,
+    name: NamedGroup::X25519MLKEM768,
+    layout: hybrid::Layout {
+        classical_share_len: X25519_LEN,
+        post_quantum_client_share_len: MLKEM768_ENCAP_LEN,
+        post_quantum_server_share_len: MLKEM768_CIPHERTEXT_LEN,
+        post_quantum_first: true,
+    },
+};
 
-impl CombinedShare {
-    fn combine(x25519: &[u8], ml_kem: kem::Ciphertext<'_>) -> Self {
-        let mut out = CombinedShare(vec![0u8; COMBINED_CIPHERTEXT_LEN]);
-        out.0[..MLKEM768_CIPHERTEXT_LEN].copy_from_slice(ml_kem.as_ref());
-        out.0[MLKEM768_CIPHERTEXT_LEN..].copy_from_slice(x25519);
-        out
-    }
-}
+/// This is the [MLKEM] key exchange.
+///
+/// [MLKEM]: https://datatracker.ietf.org/doc/draft-connolly-tls-mlkem-key-agreement
+#[allow(clippy::upper_case_acronyms)]
+pub static MLKEM768: &dyn SupportedKxGroup = &mlkem::MLKEM768;
```

Comment:
Maybe clashes between spec-native naming and Rust naming, but I feel like this would more idiomatically named `ML_KEM_768`.

Reply:
Trying to keep these named the same as the IANA names, albeit uppercased to meet the rust static/const naming convention (that doesn't affect this one, but it's why `SECP256R1` is very shouty compared to the IANA `secp256r1`).

Reply:
(realised this `clippy::upper_case_acryonyms` is not needed here; will remove)

Reply:
> Trying to keep these named the same as the IANA names, albeit uppercased to meet the rust static/const naming convention (that doesn't affect this one, but it's why `SECP256R1` is very shouty compared to the IANA `secp256r1`).

I guess that makes sense, though at the same time, I'll argue the other side: it feels to me like the Rust naming convention will be more strongly activated in the context of writing Rust code and there's little chance of confusion between `MLKEM768` and `ML_KEM_768` -- it's even likely there might be autocomplete to help out.